### DISCUSSION TITLE
[rp2040] USBH fails on some USB sticks...

### DISCUSSION THREAD

**Discussion by b-ferd** [hathach/tinyusb#2064] [Q&A]
I'm experimenting with the pico-sdk lib/tinyusb/examples/host/msc_file_explorer example.
All USB Sticks used for testing are single partition, FAT32.
Most are recognized, and the inquiry_complete_cb() function is called properly.
For some USB sticks (notably those > 8 GB, but this may be a coincidence), the inquiry_complete_cb() is never called.
Setting debuglevel to 3, the following is printed to the serial console:
```
[1:] USBH DEVICE ATTACH
Full Speed
[1:0] Open EP0 with Size = 8
Get 8 byte of Device Descriptor
[1:0] Get Descriptor: 80 06 00 01 00 00 08 00 
on EP 00 with 8 bytes
[1:0] Control FAILED, xferred_bytes = 8

[1:0] Get Descriptor: 80 06 00 01 00 00 08 00 
on EP 00 with 8 bytes
[1:0] Control FAILED, xferred_bytes = 8

[1:0] Get Descriptor: 80 06 00 01 00 00 08 00 
on EP 00 with 8 bytes
[1:0] Control FAILED, xferred_bytes = 8

[1:0] Get Descriptor: 80 06 00 01 00 00 08 00 
on EP 00 with 8 bytes
[1:0] Control FAILED, xferred_bytes = 8

[1:0:0] USBH DEVICE REMOVED
```

I'm running this on a Raspberry Pi Pico, CFG_TUH_RPI_PIO_USB set to 1, BOARD_TUH_RHPORT 0, CFG_TUH_ENABLED 1, CFG_TUH_HUB and CFG_TUH_MSC set to 1, CFG_TUH_CDC, CFG_TUH_HID and CFG_TUH_VENDOR set to 0.

Any ideas what might be going wrong here?
During my search, I stumbled across an issue here https://github.com/hathach/tinyusb/issues/1732 that seems similar, but as I understand it this is already fixed. Also I do see the multiple tries mentioned in the issue discussion, so it seems this is another issue not solved by that patch.

**Comment by b-ferd** [comment]
By the way, this is how it looks for successfully recognized stick (256 MB):

```
[1:] USBH DEVICE ATTACH
Full Speed
[1:0] Open EP0 with Size = 8
Get 8 byte of Device Descriptor
[1:0] Get Descriptor: 80 06 00 01 00 00 08 00
on EP 00 with 8 bytes
on EP 80 with 8 bytes
[1:0] Control data:
  0000:  12 01 00 02 00 00 00 40                          |.......@|
on EP 00 with 0 bytes

... (and so on)
```

**Reply by hathach** [reply]
getting host to work with all MSC devices is not an easy task. I have done this job once, each device will response differently with all kinds of expectation from host. From the log, it seems that device refuse to talk at all, since it is pio-usb, it could be a bug in the pio transportation. Please try
1. use the native host controller to see if the physical hw can enumerate the troublesome device
2. try to debug (pico debug, jlink) to see why the transfer failed. If you have an usb protocol analyzer it will help a lot.

**Reply by b-ferd** [reply]
OK will try to get the native host controller working via OTG cable. Will take a few days since I still need to order another RPI board for this to be sure I start with a clean slate.

The stick that won't work on the Rpi PIO, I cross-checked using Linux, it works there. Also did some diff analysis with lsusb and Wireshark (usbmon), so far nothing special caught my eye for the problematic stick.
The stick in question is an Intenso 16 GB, "Bus 001 Device 003: ID 058f:6387 Alcor Micro Corp. Flash Drive".
I tested the Rpi with three different sticks (two working, one not enumerated by the Rpi), and according to Linux lsusb there is no special attribute the non-working stick has the one of the other ones would not also have: max.Power is 200 mA (another one that works draws up to 900mA), bMaxPacketSize0 is 64 (the two working ones have 64 and 9), it is a USB 2.0 stick (so: 1.1 compatible), bNumEndpoints is 2 (the other working ones have 3 and 2 endpoints). 
I noticed that the first inquiry Linux sends for the Descriptor (in order to enumerate) sends this:
80 06 00 01 00 00 40 00
i.e., it requests up to 0x40 = 64 Bytes.
Tusb requests
80 06 00 01 00 00 08 00
i.e. it limits the read to 8 Bytes (just like it reports on the serial console in debug mode).
Could it be that the stick expects more bytes to be read, possibly in violation of some spec?
Or of course some subtle timing issue...

I have the RPi Debug probe (ARM SWD). Will try to look at things with a digital scope of a friend, the trusty old Tek won't cut it here...

**Comment by b-ferd** [comment]
Re-did the setup cleanly with a second RPi, still the same. Short data connections, proper power supply, 22R series resistors, but it still doesn't work the the GPIO USB (PIO with Pins 20 and 21). 
I also tried some three more USB sticks borrowed from friends, none of which worked (so it seems there are quite some sticks out there which don't work - so at least this should be easy to reproduce).

I also tried the same stick(s) with the native USB port on the RPI, using a USB-OTG cable connected to the MicroUSB-Socket, with CFG_TUH_RPI_PIO_USB set to 0 for using that port. This works with the problematic sticks!
My guess is that there is something off with the timing of the PIO state machine, but I don't have the necessary tools at hand (USB protocol analyzer, or digital scope which is fast enough and has enough storage depth) to analyze it.

Just for completeness, down below I added two log outputs, one for the working USB stick, one for the problematic USB stick, both connected to the OTG port (they both work on the OTG port, as mentioned).

_(EDIT: included log output as attachments, for readability)_
[rp2040_OTG_NokStick.txt](https://github.com/hathach/tinyusb/files/11721088/rp2040_OTG_NokStick.txt)
[rp2040_OTG_workingStick.txt](https://github.com/hathach/tinyusb/files/11721089/rp2040_OTG_workingStick.txt)

**Reply by hathach** [reply]
Long log must be attached as txt for readibility

**Reply by b-ferd** [reply]
You have a point. Sorry. Fixed.

**Reply by hathach** [reply]
maybe https://github.com/hathach/tinyusb/pull/2107 could fix this.

**Comment by b-ferd** [comment]
Switched to main/head, 13bb13204 (my last attempts were on 86c416d4c "Merge pull request #1890 from hathach/release-0.15.0", as it seems that's what's delivered with the pico-sdk).
With the updated tinyusb, now GPIO20/21 PIO usb doesn't work at all anymore, even with the previously working sticks. It does not even read the descriptor anymore or recognize the stick is connected. Hardware is unchanged. Did anything change here in the way the API should be used?
However, using RPi Pico TP3 and TP2 (i.e. native USB support) works for me as a workaround, I patched my setup to use these data lines for USB DP and DM.

Unrelated side note: it may make sense to change, in tinyusb/lib/fatfs/source/ff.h, line
`#include "ffconf.h"`
to
`#include <ffconf.h>`
This would allow to override the fatfs config with your own includefile, by specifying a local include path that has priority over the fatfs one. I know this is a fatfas issue, so not sure if that's on your plate. Just thought I'd mention it. Not having to edit the lib files in place would be much cleaner.

**Reply by hathach** [reply]
pio-usb pin changes to to DP 16, 17 to match adafruit feather usb host. 

ffconfig.h can be moved to example folder, may do it later when I have time
